src/events/upgrade.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handler for handling OpensearchDashboards in-place upgrades."""
+import logging
+import time
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import (
+    BaseModel,  # pyright: ignore [reportGeneralTypeIssues, reportAttributeAccessIssue]
+)
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    UpgradeGrantedEvent,
+)
+from typing_extensions import override
+
+if TYPE_CHECKING:
+    from charm import (
+        OpensearchDashboardsCharm,  # pyright: ignore [reportGeneralTypeIssues, reportAttributeAccessIssue]
+    )
+
+logger = logging.getLogger(__name__)
+
+
+class OpensearchDashboardsDependencyModel(BaseModel):
+    """Model for OpensearchDashboards Operator dependencies."""
+
+    osd_upstream: DependencyModel
+
+
+class ODUpgradeEvents(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "OpensearchDashboardsCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+    def post_upgrade_check(self) -> None:
+        """Runs necessary checks validating the unit is in a healthy state after upgrade."""
+        self.pre_upgrade_check()
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        pass
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
```

Comment:
if we are following the process as OpenSearch, this is not needed anymore.

Reply:
As for now, we are not following the Opensearch solution It has two reasons.

Within this code I have "inherited" a full solution for v1 -- which makes this change really a low-hanging fruit ;-)
Moving from a functional 'upgades v1' is to 'v2' is easier than (re-)implementing all from scratch. Not to mention that v1 already provides a fully functional upgrades feature to the Dashboards.

In the meanwhile, we may not even need v2 for the Dashboards, as --after all-- we are dealing with a lightweight, stateless charm here. 

Alltogether: it's either a simple temporary change delivering quite some value, or may even be the permanent solution.